src/main/java/org/opensearch/flowframework/workflow/WorkflowProcessSorter.java

```diff
@@ -244,11 +245,32 @@ public void validateGraph(List<ProcessNode> processNodes, WorkflowValidator vali
                 );
             }
         }
+    }
 
+    private WorkflowValidator readWorkflowValidator(String workflowId) {
+        try {
+            return WorkflowValidator.parse("mappings/workflow-steps.json");
+        } catch (Exception e) {
+            logger.error("Failed to read workflow-steps mapping file", e);
+            throw new FlowFrameworkException(
+                "Workflow " + workflowId + " failed at reading workflow-steps mapping file",
+                RestStatus.INTERNAL_SERVER_ERROR
+            );
+        }
     }
 
-    private TimeValue parseTimeout(WorkflowNode node) {
-        String timeoutValue = (String) node.userInputs().getOrDefault(NODE_TIMEOUT_FIELD, NODE_TIMEOUT_DEFAULT_VALUE);
+    /**
+     * A method for parsing workflow timeout value.
+     * The value could be parsed from node NODE_TIMEOUT_FIELD, the timeout field in workflow-step.json,
+     * or the default NODE_TIMEOUT_DEFAULT_VALUE
+     * @param node the workflow nde
+     * @return the timeout value
+     */
+    protected TimeValue parseTimeout(WorkflowNode node) {
+        WorkflowValidator validator = readWorkflowValidator(node.id());
+        String nodeTimeoutValue = Optional.ofNullable(validator.getWorkflowStepValidators().get(node.type()).getTimeout())
+            .orElse(NODE_TIMEOUT_DEFAULT_VALUE);
+        String timeoutValue = (String) node.userInputs().getOrDefault(NODE_TIMEOUT_FIELD, nodeTimeoutValue);
         String fieldName = String.join(".", node.id(), USER_INPUTS_FIELD, NODE_TIMEOUT_FIELD);
         TimeValue timeValue = TimeValue.parseTimeValue(timeoutValue, fieldName);
         if (timeValue.millis() < 0) {
```

Comment:
I'm OK with requiring positive values.  Note that the `TimeValue` class does provide an explicit exception for -1 as a 'magic number' which could potentially be allowed as a special-case "no timeout" value (if documented).

However, as strongly suggested elsewhere I think this parsing needs to be moved to the file reading code and properly unit tested.